### ISSUE TITLE
feat: abstract patch options

### ISSUE THREAD

**Issue by oSumAtrIX** [ReVanced/revanced-patcher#163]
### Type

Functionality

### Issue

ReVanced Patcher should only provide primitive types for patch options such as an `IntOption` or `StringOption`. This creates complications when abstractions are required at a higher level such as when a patch has a color option. ReVanced Manager for example can not differentiate between colors and strings. 

### Feature

Allow options to be extendable. Patches can provide their own sets of abstractions for `PatchOption` such as a `ColorOption` inheriting from the primitive `StringOption`. These options should be in a separate module, published as a maven library to be used by higher-level applications such as ReVanced Manager. The patches would use the abstract options and shade them at compile time. Higher-level applications would use the library to compile against. ReVanced Manager could then use `ColorOption` to display a color picker. Upon loading the patches, `ClassLoader` will be aware of the abstract options at runtime. If the higher-level application is not aware of the abstract options, such as when not compiling against the abstract options library, then it will treat the options as the primitive option they inherit from. This is the instance when a patch, for example, does not make use of our abstract options but for example their own. ReVanced Manager would not be aware of these options and thus fall back to the primitive option unless ReVanced Manager is compiled against these abstract options. By default, ReVanced high-level applications like ReVanced Manager should compile against ReVanced abstract options as a base abstraction. Other patches outside of ReVanced are free to use the ReVanced library or extend the primitive options on their own, with the limitation, of high-level applications falling back to the primitive type.

TLDR:

Patches provide abstract options such as `ColorOption` by inheriting from ReVanced Patcher primitive options. High-level applications can be aware of those options by compiling against them. If they are unaware of the abstract options, they can fall back to the primitive type the option inherits from.

### Motivation

This change would come in favor of a higher level for UX. A color picker is more favorable than an input box with the necessity of the user having to infer the real type of the option such as when a text box is provided but a hex color is expected.

### Acknowledgements

- [X] I have searched the existing issues and this is a new and no duplicate or related to another open issue.
- [X] I have written a short but informative title.
- [X] I filled out all of the requested information in this issue properly.
